Review Snake 1.1 lizavetashap

lizavetashap opened the pull request:
Сделала большой первую букву в слове Apple

Commit:
Apple

Commit:
Hangman - done

Commit:
Snake - demo
иногда ломается и растет бесконечно, не убивается об себя (в коде закомменчена возможность, не знаю будет ли она работать). из приятного - ест яблоко, убивается об стенки, чаще всего нормально растет и управляется

lizavetashap commented:
комментарий к змейке: иногда ломается и растет бесконечно, не убивается об себя (в коде закомменчена возможность, не знаю будет ли она работать). из приятного - ест яблоко, убивается об стенки, чаще всего нормально растет и управляется

lizavetashap commented:
@doreshnikov

Commit:
Snake - done
работает:
убивается об стены и себя
больше не растет бесконечно
ест яблоко и растет на один
с помощью Esc можно завершить игру

lizavetashap commented:
комментарий к змейке 2, работает:
убивается об стены и себя
больше не растет бесконечно
ест яблоко и растет на один
с помощью Esc можно завершить игру

+ почищен синтаксис

doreshnikov commented on .idea/misc.xml:
Такие файлы лучше изначально в .gitignore добавлять, чтобы они в репозитории не появлялись :)
Сейчас уже не поможет, но так, на будущее

doreshnikov commented on hang-snake/hangman/run.py:
```diff
@@ -1,33 +1,131 @@
+    return 'солнышко'
+word = create_secret()
+
+FIELD = (
```
Жесть, тут еще и виселица строится поэлементно...
Ну почему бы и нет.
Но чтобы оно в коде, в котором логика действий описывается, много места не занимало, имеет смысл в отдельный файл вынести.

doreshnikov commented on hang-snake/hangman/run.py:
```diff
@@ -1,33 +1,131 @@
+    """
+)
+number_of_errors = 0 #кол-во ошибок сделанных игроком
+maximum_erorrs = 10 #максимальное количество ошибок
```
Константы обычно пишутся капсом, но это мелочи

doreshnikov commented on hang-snake/snake/run.py:
```diff
@@ -1,36 +1,101 @@
+# наше поле
+def print_field():
+    for cell in CELLS:
+        if cell in snake:
```
Эта проверка работает за время, пропорциональное длине змейки в среднем.
Имеет смысл сделать поле в виде таблицы символов, и потом просто пройтись по змейке и для каждого ее элемента, поставить нужный символ в соответствующую ячейку.

doreshnikov commented on hang-snake/snake/run.py:
```diff
@@ -1,36 +1,101 @@
+    head_snake = snake[0][0] + direction[0], snake[0][1] + direction[1]
+    # делаем змейке новую голову, чтобы она ползла вперед
+    snake.insert(0, head_snake)
+    if not eat: # если змея не съела яблоко то удаляем хвост
```
Если уж эта логика вынесена в функцию, то логично целиком логику в нее внести.
Уж точно не делать `global eat`, а сделать хотя бы `eat` параметром функции - так хотя бы поведение функции не будет зависеть от "внешних" глобальных параметров. Типа `def snake_move(eat: bool): ...`

Ну и поскольку к концу ее выполнения всегда `eat == False`, то можно вообще `eat` снаружи не хранить, а просто передать в функцию факт "было ли съедено яблоко".

doreshnikov reviewed (commented):
Вроде все выглядит клево, чуть более подробный ревью сделаю, когда классы появятся :)

Commit:
проверка сохранение промежуточного

Commit:
проверка сохранение промежуточного 2

Commit:
snake-hangman
змейка и виселица обернуты в классы.
виселица работает от генератора букв в файле common
в змейку генерируется буква из common и пока змея ее не съест - буква не поменяется
как "сделать поле в виде таблицы символов, и потом просто пройтись по змейке и для каждого ее элемента, поставить нужный символ в соответствующую ячейку" - пока не поняла, надеюсь поможете на практике
переменную eat не внесла в функцию, потому что она используется не только в ней, поэтому нам нужно чтобы она была в __init__ (насколько я поняла ваш комментарий)
остальное вроде сделала

lizavetashap commented:
@doreshnikov

lizavetashap commented:
snake-hangman
змейка и виселица обернуты в классы.
виселица работает от генератора букв в файле common
в змейку генерируется буква из common и пока змея ее не съест - буква не поменяется
как "сделать поле в виде таблицы символов, и потом просто пройтись по змейке и для каждого ее элемента, поставить нужный символ в соответствующую ячейку" - пока не поняла, надеюсь поможете на практике
переменную eat не внесла в функцию, потому что она используется не только в ней, поэтому нам нужно чтобы она была в __init__ (насколько я поняла ваш комментарий)
остальное вроде сделала

Commit:
snake-hangman demo связь
это проверять пока не надо, тут идет работа, все сырое

Commit:
snake-hangman demo2 связь
это проверять пока не надо, тут идет работа, все сырое

Commit:
snake-hangman demo3 связь
это проверять пока не надо, тут идет работа, все сырое
помогите только если увидите с ссылкой файлов друг на друга, не понимаю, как в main подать в провайдер хэнгмена букву из снэйка

Commit:
snake-hangman demo4
ну это в общем последнее чего я пока добилась

doreshnikov commented on hang-snake/common/providers.py:
```diff
@@ -0,0 +1,27 @@
+
+class SecretLetterProvider(LetterProvider):
+    def get_next_letter(self) -> str:
+        return create_secret()[random.randint(0, len(create_secret())-1)]
```
Хорошая идея :)
Но вообще в идеале `create_secret` должен каждый раз что-то новое возвращать
Так что лучше в конструкторе к этому провайдеру передавать то загаданное слово, которое уже было выбрано

doreshnikov commented on hang-snake/hangman/run.py:
```diff
@@ -1,33 +1,71 @@
-SECRET = create_secret()
-n = len(SECRET)
+    def read_guess(self, letter) -> str:
+        while True:
```
Это выглядит странно. У тебя же `read_guess` вызывается снаружи. Если тебе передали некорректную букву, у тебя этот цикл будет бесконечным

doreshnikov commented on hang-snake/common/main.py:
```diff
@@ -0,0 +1,64 @@
+from snake.run import Snakegame
+import  time
+
+s_game = Snakegame(RandomLetterProvider(), SecretLetterProvider())
```
Судя по всему, игры пока не особо соединены.
Вообще тебе бы в идеале сделать еще один провайдер типа `QueueProvider` (с очередью). И в него должно быть можно класть буквы из змейки, а доставать - в виселице. Тогда будет прям то, что нужно.

UPD: прочитал твои правила, может можно и без этого вообще))

doreshnikov commented on hang-snake/common/rules.py:
```diff
@@ -0,0 +1,17 @@
+         ' guessed right, the letter will appear in the word\n 4. after each choice of letters, you will see how close'
+         ' the little man is to death\n 5. try not to kill your ward and guess the word.\n Good luck)\n\n'
+         
+         'ПРАВИЛА ИГРЫ\n 1. на экране будут появляться 5 букв. Вы можете собрать 3, после этого буквы поменяются.\n'
```
Правила клевые, мне нравится)

doreshnikov commented on hang-snake/hangman/run.py:
```diff
@@ -1,33 +1,71 @@
+        return self.MAX_ERRORS == self.number_of_errors
+
+    def step(self):
+        letter = self.read_guess()
```
Вот тут как будто бы не работает... У тебя `read_guess` ожидает, что ты ему букву передашь. Непонятно, почему, и откуда ты ее возьмешь..

doreshnikov commented on hang-snake/snake/run.py:
```diff
@@ -1,36 +1,156 @@
+        return matrix
+
+    def print_snake_and_apples(self, matrix):
+        for y, row in enumerate(matrix):
```
Про это я уже писал, попробую подробнее объяснить. Чтобы быстрее строилось, можно завести массив `matrix: list[list[str]]`, то есть массив массивов символов - и его заполнять, типа
```py
for x, y in self.snake:
    matrix[x][y] = '■'
```

Ну и так далее с другими объектами

doreshnikov commented on hang-snake/snake/run.py:
```diff
@@ -1,36 +1,156 @@
+            print(wrong_input)
+
+    def give_letter_to_provider(self):
+        return self.index_letter
```
Разве здесь должен быть не `self.chosen_letter`?

doreshnikov reviewed (commented):
Я поскольку не успел пока у себя позапускать, как будто есть какие-то моменты, которые на взгляд пока не работают, но, думаю, на практике разберемся

Commit:
snake-hangman demo5
у меня не получилось сделать тот выод, который я хотела, поэтому пришлось изменить вывод поля в виселице и сделать мультипринтер
виселица окончательно работает, змейка в процессе редакции и добавления вывода через принтер

Commit:
snake-hangman demo6
вроде змейка переделана под что-то рабочее, но связь пока не доработана.

Commit:
snake fixes
змейка окончательно переделано под рабочую с готовой передачей в виселицу и правильной генерацией букв. все короче там работает, но связь пока не доработана.

Commit:
snake-hang 1.0
все работает. букву ввести можно, но пока что ожидание и все устройство инпута работает не очень красиво (то есть тяжело что-то ввести, потому что поле постоянно обновляется, но это НЕ невозможно и передает правильную букву и т.д.)

lizavetashap commented:
все работает. букву ввести можно, но пока что ожидание и все устройство инпута работает не очень красиво (то есть тяжело что-то ввести, потому что поле постоянно обновляется, но это НЕ невозможно и передает правильную букву и т.д.)

Commit:
snake-hang 1.1
все работает. буква вводится нормально (иногда со второго раза, но это мелочи)

lizavetashap commented:
все работает. букву ввести можно, убрано почти мерцание, ввод адекватный
+добавлена проверка на повторяющиеся выбранные буквы
+ добавлен список слов из которых вылетает рандомное

lizavetashap commented:
@doreshnikov
